flink-autoscaler-standalone/src/main/java/org/apache/flink/autoscaler/standalone/StandaloneAutoscalerExecutor.java

```diff
@@ -48,22 +59,34 @@ public class StandaloneAutoscalerExecutor<KEY, Context extends JobAutoScalerCont
     private final AutoScalerEventHandler<KEY, Context> eventHandler;
     private final JobAutoScaler<KEY, Context> autoScaler;
     private final ScheduledExecutorService scheduledExecutorService;
+    private final ExecutorService scalingThreadPool;
 
     public StandaloneAutoscalerExecutor(
-            @Nonnull Duration scalingInterval,
+            @Nonnull Configuration conf,
             @Nonnull JobListFetcher<KEY, Context> jobListFetcher,
             @Nonnull AutoScalerEventHandler<KEY, Context> eventHandler,
             @Nonnull JobAutoScaler<KEY, Context> autoScaler) {
-        this.scalingInterval = scalingInterval;
+        this.scalingInterval = conf.get(CONTROL_LOOP_INTERVAL);
         this.jobListFetcher = jobListFetcher;
         this.eventHandler = eventHandler;
         this.autoScaler = autoScaler;
         this.scheduledExecutorService =
                 Executors.newSingleThreadScheduledExecutor(
                         new ThreadFactoryBuilder()
-                                .setNameFormat("StandaloneAutoscalerControlLoop")
+                                .setNameFormat("autoscaler-standalone-control-loop")
                                 .setDaemon(false)
                                 .build());
+
+        int parallelism = conf.get(CONTROL_LOOP_PARALLELISM);
+        this.scalingThreadPool =
+                new ThreadPoolExecutor(
+                        parallelism,
+                        parallelism,
+                        0L,
+                        TimeUnit.MILLISECONDS,
+                        new LinkedBlockingQueue<>(parallelism * 4),
```

Comment:
Why `*4`? Shouldn't `parallelism` be enough for the work queue?

Reply:
First of all, why this `scalingThreadPool` doesn't use unbounded `LinkedBlockingQueue`?

My thought is if all threads are busy and `LinkedBlockingQueue` has a lot of tasks, we can execute `task` in control loop scheduler thread. So I choose the `new ThreadPoolExecutor.CallerRunsPolicy()`.

> Why *4?

It's not strict. My thought is scheduler thread only work when `LinkedBlockingQueue` has a lot of tasks. So parallelism * 3 or > 3 makes sense.

-------------------------------------------

Of course, if you think `scalingThreadPool` already has a lot of thread, and we don't need thejcontrol loop scheduler thread to execute any tasks. We can let control loop scheduler thread only submit tasks, and wait for all tasks are finished. It means we don't let control loop scheduler thread executes tasks.

WDYT?

Reply:
What is the benefit of the queue? If we are waiting on all tasks to finish, then we do not need the queue.

I think the current logic of waiting for all tasks to finish is a bit fragile. For example, when we are executing a scaling decision, we may block for a long time on savepointing which will halt the next scaling for all pipelines. We could alternatively not block all but only the specific job for the next scaling.

Reply:
> What is the benefit of the queue? If we are waiting on all tasks to finish, then we do not need the queue.

Sounds make sense, if we want to waiting, the queue isn't useful.

> I think the current logic of waiting for all tasks to finish is a bit fragile. For example, when we are executing a scaling decision, we may block for a long time on savepointing which will halt the next scaling for all pipelines. We could alternatively not block all but only the specific job for the next scaling.

Sorry, I didn't notice the `scalingSingleJob` will cost too many times. Currently, the scaling action only call the rescale api, so the thread won't be wait the savepointing.

But I think your concern is reasonable. We define the `ScalingRealizer` interface, it means users can custom define the action. It might cost long time.

--------------------------------------------------------


Based on your suggestion, I think the `control loop scheduler thread` cannot wait all scaling. When the scaling of one job is slow, it shouldn't affect other jobs.

Also, if a new scaling round is started, and the last scaling of some of jobs aren't finished. Should we ignore this round for these jobs? Some reasons might cause this issues:

- The scaling action is slow
- It has too many jobs, and the thread number isn't enough. So the last round always cannot be finished when the new round is started.

So, I think we should ignore unfinished jobs in the new round.

WDYT?

Reply:
>Based on your suggestion, I think the control loop scheduler thread cannot wait all scaling. When the scaling of one job is slow, it shouldn't affect other jobs.

Yes, I agree.

>Also, if a new scaling round is started, and the last scaling of some of jobs aren't finished. Should we ignore this round for these jobs?

Yes, that would be great. We just wait until they are ready. In the Kubernetes operator, there is currently no timeout for single jobs but we could add one in the future if this poses a problem.